### DISCUSSION TITLE
Playing engines (without Centaur Adaptive Mode)

### DISCUSSION THREAD

**Discussion by EdNekebno** [EdNekebno/DGTCentaurMods#56] [Show and tell]
gamemanager.py is far enough along now that it can be used for some interesting things. One of the simplest is to query a uci chess engine.

I've added a script stockfish.py that calls stockfish directly. 
I've added an additional folder called engines in which to place other engines we might want to use. I was told about CT800 and I quite like it (it can play lower ELO so it suits me :) ), so I've added that. I also created ct800.py to call it.

I've put these on the menu. They might not be 100% (I had some problems with pawn promotion) but I think they're worth having there whilst I keep working. Playing CT800 at a lower level was a whole new experience with the board for me.

(also in that push was a couple of bug fixes. The most important that in some cases moving a piece to a1 wasn't detected...because...well, maths. -0 == +0 )

**Comment by wormstein** [comment]
This is a very cool feature and it's a thing I was thinking about it too - have the board work as a GUI and use it with any engine that supports UCI.
I would say not to add any engine binary to our repo so we keep it slim. I could build a sh script to grab and compile engines as needed and integrate that into the setup process... Just put an empty file with the same name as a placehholder in engines folder.

CT is fine, I see it;s only 200 KB.

**Comment by EdNekebno** [comment]
Yeh...it's tiny. Maybe no more though. One of the things I was thinking we can do in the future is make this a bit more universal. If we can do a script that iterates the engines folder for engines and menu them and the uci script worked generically  - then we could let people upload an engine via the web interface. End users can then have their engines of choice on the board. 

But that's more a future thought. For now CT800 makes a world of difference. I promise I won't upload any more. Well, unless I find a kind of uci proxy engine that sends the moves over the network for somebody/something else to answer. But that's the only exception, honest. And I think I'm unlikely to find it :)

**Comment by wormstein** [comment]
I think we can add the engines folder to ,gitidnore as well.

Addidng the engines in the menu sound like an easy task for not to iterate engines like you said. It will become easier once we have the database backend so when an user adds the engine he would attach a name to it which we can read from database and add a menu item.... so it looks good :) Well, this sounds like a first table.

Also I think we can redo the menu. Have an option called: "Play!" in which we can add DGT Centaur, Lichess online, and the rest of engines....

**Comment by EdNekebno** [comment]
Sounds like a plan. I like the "Play!" idea. One thing to keep in mind with the menu as it is is that scrolling up/down is faster, going to a separate submenu is quite slow as it refreshes the whole screen. I'm not sure if python does lists of lists but something I'm wondering is if there's a way we can build a structure of menu information and have all the code for each item that calls a script rather than a submenu in functions and the list structure to contain the function to call. Then rather than drawing a new page it can update the same one and should be quicker. Also better if we eventually move the menu structure into a database. Database menu could prove very useful as different engines also have different options and the higher level ELO modes for me are a waste of menu space :)

It feels like I got a new board, dumbed down to my level :)

**Reply by wormstein** [reply]
A json structure might work to store the menu infotmation. I need to refresh my memory on how the structure works, but there are some online builders. Here is one: http://www.objgen.com/json?demo=true The demo is very helpful.

Something like:
``` json
{
  "Play": {
    "DGT": {},
    "e-Board": {},
    "Lichess": {}
  },
  "Settings": {}
}
```
Json can let us add keys for each item, and then all menu can be generated by a singlue function only. Search the web, there are some examples on using json for a menu structure...
I hope this helps. 

There are so many things we could focus now, I'd say we should make a priority list. For example, the menu is good enough for now. I am focusing on having a setup process and a first release. We could branch for a new set of features. We need to add new features instead of rebuilding ones that work.
Once I am confident the building scripts are good to go, I let you know and we decide when to tag the repo for a release and give it a go. I think we have enough things onboard for a v1.0 already :))

**Reply by EdNekebno** [reply]
Very close...great minds. I'm trying list of lists and it's very fast. e.g.

```
menu = [
    ['Centaur',centaurfunc,None],
    ['Lichess',None,None],
    ['Engines',None,
        [
            ['Stockfish',None,[
                ['Pure',None,None],
                ['1350',None,None]
            ]],
            ['CT800',None,None]
        ]
    ],
    ['e-Board',None,None],
    ['Chromecast',None,None],
    ['Settings',None,None],
    ['Get Support',None,None]
]
```

In each list the second parameter is a function to call that gets passed the menu hierarchy to get there - or None if there isn't one. The third parameter is any submenu. If I can iron out the bugs it'll go fast in comparison to what we had as the epaper has the whole menu at once for updates.

**Reply by EdNekebno** [reply]
You're right actually, I'd rather resolve any bugs that show in gamemanager and have things like lichess running on gamemanager so I can be more confident in the lichess play and add features. Improved menu can wait.

For features - I think it might be time to action the database plan. You were going to use SqlAlchemy right? Then it looks like we just use pip to install and pick a location for our database right? https://towardsdatascience.com/sqlalchemy-python-tutorial-79a577141a91  or is there more to it than that? I need to play plenty of test games to iron out any bugs in what we have - and it may as well then be recording the games to a DB to generate PGN files later (not sure why I hadn't added PGN to the Todo list)

A release will be exciting. Maybe we should put the word beta in there or something!

**Comment by EdNekebno** [comment]
I'm almost certainly going to change the way the menu works for the millionth time by the way. I think I've found a much quicker way.
